File: 75-sort-colors/75-sort-colors.rb
# @param {Integer[]} nums
# @return {Void} Do not return anything, modify nums in-place instead.
def sort_colors(nums)
    l, m, r = 0, 0, nums.size - 1
    until m > r do 
        if nums[m] == 0
            swap nums, l, m
            l += 1
        elsif nums[m] == 2
            swap nums, m, r
            r -= 1
            m -= 1
        end
        m += 1
    end
end

def swap(nums,l,r)
    nums[l], nums[r] = nums[r], nums[l] 
end

File: medium/75-sort-colors/75-sort-colors.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
// with bucket sort: O(n) time & space
// with 2 pointers => O(n) time, O(1) space

const sortColors = function(nums) {
    let l = 0,
        r = nums.length - 1,
        i = 0;
    const swap = (i, j) => {
        [nums[i], nums[j]] = [nums[j], nums[i]];
    }
    
    while (i <= r) {
        if (nums[i] === 0) {
            swap(l, i);
            l++;
        }
        else if (nums[i] === 2) {
            swap(i, r);
            r--;
            // dont want to increment i here
            i--;
        }
        i++;
    }
};

